sdk/communication/azure-communication-callautomation/azure/communication/callautomation/_shared/models.py

```diff
@@ -72,74 +76,76 @@ class CommunicationIdentifier(Protocol):
+    """The type of identifier."""
+    properties: CommunicationUserProperties
+    """The properties of the identifier."""
+    raw_id: str
+    """The raw ID of the identifier."""
 
     def __init__(self, id: str, **kwargs: Any) -> None:
-        self.raw_id = kwargs.get("raw_id", id)
+        """
+        :param str id: ID of the Communication user as returned from Azure Communication Identity.
+        :keyword str raw_id: The raw ID of the identifier. If not specified, the 'id' value will be used.
+        """
         self.properties = CommunicationUserProperties(id=id)
-        if self.raw_id is None:
-            self.raw_id = _communication_user_raw_id(self)
+        self.raw_id = kwargs.get("raw_id") or id
 
     def __eq__(self, other):
         try:
-            return self.raw_id == other.properties["id"]
-        except Exception:  # pylint: disable=broad-except
+            return self.raw_id == _communication_user_raw_id(other)
+        except (AttributeError, TypeError, IndexError):
             return False
 
 
 def _communication_user_raw_id(identifier: CommunicationUserIdentifier) -> str:
-    if identifier.raw_id:
-        return str(identifier.raw_id)
-    return str(identifier.properties["id"])
+    if hasattr(identifier, "raw_id"):
+        return identifier.raw_id
+    return identifier.properties["id"]
 
 
-PhoneNumberProperties = TypedDict("PhoneNumberProperties", {"value": str})
+class PhoneNumberProperties(TypedDict):
+    """Dictionary of properties for a PhoneNumberIdentifier."""
+    value: str
+    """The phone number in E.164 format."""
 
 
 class PhoneNumberIdentifier:
-    """Represents a phone number.
-
-    :ivar str raw_id: Optional raw ID of the identifier.
-    :ivar kind: The type of identifier.
-    :vartype kind: str or CommunicationIdentifierKind
-    :ivar Mapping properties: The properties of the identifier.
-     The keys in this mapping include:
-        - `value`(str): The phone number in E.164 format.
-
-    :param str value: The phone number.
-    """
-
-    kind = CommunicationIdentifierKind.PHONE_NUMBER
+    """Represents a phone number."""
+    kind: Literal[CommunicationIdentifierKind.PHONE_NUMBER] = CommunicationIdentifierKind.PHONE_NUMBER
+    """The type of identifier."""
+    properties: PhoneNumberProperties
+    """The properties of the identifier."""
+    raw_id: str
+    """The raw ID of the identifier."""
 
     def __init__(self, value: str, **kwargs: Any) -> None:
-        self.raw_id = kwargs.get("raw_id")
+        """
+        :param str value: The phone number.
+        :keyword str raw_id: The raw ID of the identifier. If not specified, this will be constructed from
+          the 'value' parameter.
+        """
         self.properties = PhoneNumberProperties(value=value)
-        if self.raw_id is None:
-            self.raw_id = _phone_number_raw_id(self)
+        self.raw_id = kwargs.get("raw_id") or _phone_number_raw_id(self)
 
     def __eq__(self, other):
```

Comment:
Not sure why we removed the typing from here?

Reply:
I think it was just refactoring, originally we would set raw_id like this:

    _self.raw_id = kwargs.get("raw_id")
    if self.raw_id is None:
                self.raw_id = _phone_number_raw_id(self)_
		
Which we changed to the following (which should be equivalent, hopefully):
    _self.raw_id = kwargs.get("raw_id") or _phone_number_raw_id(self)_

Reply:
However, I reverted to old way, because then the raw_id property was missing in the call to _phone_number_raw_id(self), which forced me to use the hasattr() function, but as pointed out, that slightly changes the behavior on equality check, so I reverted.

Reply:
Aaah thank you for this!
I was actually referring to the type hints for the `__eq__` method, however this description has helped me understand the other problem - let me add a fix for this....